File: Data_Structures/015_vectors/vectors.java
// util vession

import java.util.LinkedList;
import java.util.Vector;

class VECTOR{
  // private
  Vector<Integer> vec = new Vector<Integer>();

  // constructor
  public Vector<Integer> Vector(){
    return vec;
  }

  public void setSize(int n){
    vec.setSize(n);
  }

  // public
  public int getLength(){
    return vec.size();
  }

  public void store(int pos, int val){
    vec.set(pos, val);
  }

  public int select(int pos){
    return vec.get(pos);
  }
}

class MAIN{
  // main
  static public void main(String[] args){
    /* example in video */
    VECTOR v1 = new VECTOR();
    v1.setSize(4);
  
    for(int i = 0; i < v1.getLength(); i++)
      v1.store(i, i+1);

    for(int i = 0; i < v1.getLength(); i++)
      System.out.print(v1.select(i) + " ");
    System.out.println("");

    /* review question 6 */
    VECTOR v2 = new VECTOR();
    v2.setSize(4);

    v2.store(0, 2);

    for(int i = 1; i < v2.getLength(); i++)
      v2.store(i, (i + 1) * v2.select(i - 1));

    System.out.println("x : " + v2.select(3));
  }
}

File: Data_Structures/015_vectors/vectors.js
// vector is fixed-size array
// as JavaScript array is resizeable, need a function constructor to limit actios of array

class Vector {
  // private
  #vec = [];

  // public
  constructor(length) {
    this.#vec = new Array(length);
  }

  getLength = function () {
    return this.#vec.length;
  };

  store = function (pos, val) {
    if (pos < 4) this.#vec[pos] = val;
  };

  select = function (pos) {
    return this.#vec[pos];
  };
}

/* example in video */
const v = new Vector(4);

// store values into vector
for (let i = 0; i < v.getLength(); i++) v.store(i, i + 1);
// get length of the vector
console.log(v.getLength());
// get values from the vector
for (let i = 0; i < v.getLength(); i++) console.log(v.select(i));

/* review question 6 */
const w = new Vector(4);

w.store(0, 2);

for (let i = 1; i < w.getLength(); i++) w.store(i, (i + 1) * w.select(i - 1));

console.table({ x: w.select(3) });
